working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,416 @@
+}
+```
+
+But that code requires iterating over the fields of the class somehow. Short of
+using mirrors or code generation, there’s no way to write a reusable system that
+can add JSON serialization/deserialization to any given class.
+
+### Data classes
+
+The [most requested open language issue][data_classes_issue] is to add data
+classes. A data class is essentially a regular Dart class that comes with an
+automatically provided constructor and implementations of `==`, `hashCode`, and
+`copyWith()` (called `copy()` [in Kotlin][kotlin_copy]) methods based on the
+fields the user declares in the class.
+
+The reason this is a *language* feature request is because there’s no way for a
+Dart library or framework to add data classes as a reusable mechanism. Again,
+this is because there isn’t any easily available abstraction that lets a Dart
+user express “given this set of fields, add these methods to the class”. The
+`copyWith()` method is particularly challenging because it’s not just the *body*
+of that method that depends on the surrounding class’s fields. The parameter
+list itself does too.
+
+We could add data classes to the language, but that only satisfies users who
+want a nice syntax for *that specific set of policies*. What happens when users
+instead want a nice notation for classes that are deeply immutable,
+dependency-injected, [observable][observable], or
+[differentiable][differentiable]? Sufficiently powerful static metaprogramming
+could let users define these policies in reusable abstractions and keep the
+slower-moving Dart language out of the fast-moving methodology business.
+
+### Flutter verbosity
+
+To get better incremental update performance and more maintainable fine-grained
+code, it’s often useful in Flutter to break a large widget with a complex
+`build()` method down into an aggregation of smaller widget classes. But
+hoisting some layout code into a new widget class requires enough boilerplate
+that users avoid it. For example, say you have a widget with some fixed data,
+stateful data, and a bit of build code that you want to pull into a new widget
+class `Foo`. You have to write:
+
+```dart
+class Foo extends StatefulWidget {
+  Foo({ Key key, this.fixedStuff }) : super(key: key);
+  FixedStuff fixedStuff;
+  State<Foo> createState() => _FooState();
+}
+
+class _FooState extends State<Foo> {
+  StatefulStuff statefulStuff = ...;
+  @override
+  Widget build(BuildContext context) {
+    buildCode();
+  }
+}
+```
+
+The widget name is repeated six times, `fixedStuff` appears twice, and the
+remaining code is almost pure boilerplate that is identical across all
+`StatefulWidget` and `State` subclasses.
+
+There are other patterns in Flutter like this that are commonly applied to
+widget classes but require Flutter users to write a lot of repetitive code.
+Metaprogramming could enable the Flutter framework to express these patterns
+more tersely and with less chance for user error.
+
+## Requirements
```

Comment:
What about configuration? What inputs will a code generator have access to? Only the input (Dart) file?

Can they read other Dart files? Other (configuration) files? Yaml?

Reply:
This is all TBD